File: frontend/src/lib/seedream.ts
interface SeedreamRequest {
  prompt: string
  image?: string[]
  size?: string
  response_format?: 'url' | 'b64_json'
  output_format?: 'png' | 'jpeg'
  watermark?: boolean
  n?: number
}

interface SeedreamDataItem {
  url?: string
  b64_json?: string
  size?: string
}

interface SeedreamApiResponse {
  model: string
  created: number
  data: SeedreamDataItem[]
  usage: {
    generated_images: number
    output_tokens: number
    total_tokens: number
  }
}

export async function callSeedream(
  params: SeedreamRequest
): Promise<{ images: string[] }> {
  const apiKey = process.env.SEEDREAM_API_KEY
  const apiUrl = process.env.SEEDREAM_API_URL
  const model = process.env.SEEDREAM_MODEL

  if (!apiKey || !apiUrl || !model) {
    throw new Error('Seedream API configuration missing')
  }

  const body: Record<string, unknown> = {
    model,
    prompt: params.prompt,
    size: params.size ?? '2K',
    response_format: params.response_format ?? 'url',
    output_format: params.output_format ?? 'jpeg',
    watermark: params.watermark ?? false,
    n: params.n ?? 1,
  }

  if (params.image && params.image.length > 0) {
    body.image = params.image
  }

  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(body),
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Seedream API error: ${response.status} - ${error}`)
  }

  const result: SeedreamApiResponse = await response.json()

  const images = result.data
    .map((item) => item.url ?? item.b64_json ?? '')
    .filter(Boolean)

  if (images.length === 0) {
    throw new Error('Seedream returned no images')
  }

  return { images }
}

File: backend/app/services/seedream.py
from openai import OpenAI

from app.config import settings

_client: OpenAI | None = None


def _get_client() -> OpenAI:
    global _client
    if _client is None:
        _client = OpenAI(
            api_key=settings.seedream_api_key,
            base_url=settings.seedream_base_url,
            timeout=120.0,
        )
    return _client


def call_seedream(
    prompt: str,
    images: list[str] | None = None,
    size: str = "2K",
    output_format: str = "jpeg",
    n: int = 1,
) -> list[str]:
    client = _get_client()

    extra_body: dict = {
        "output_format": output_format,
        "watermark": False,
    }
    if images:
        extra_body["image"] = images

    response = client.images.generate(
        model=settings.seedream_model,
        prompt=prompt,
        size=size,  # type: ignore[arg-type]
        response_format="url",
        n=n,
        extra_body=extra_body,
    )

    result_images = [
        item.url or (item.b64_json or "")
        for item in response.data
    ]
    result_images = [img for img in result_images if img]

    if not result_images:
        raise ValueError("Seedream returned no images")

    return result_images
